### PR TITLE
Add end-to-end `heartbeat_api` event

### PR DESCRIPTION
**Pull request by Twixes**
Adds optional `heartbeat_api` event, allowing full end-to-end measurement of ingestion lag.

Config

### CODE REVIEW

File: plugin.json [hunk 1]
```diff
@@ -2,10 +2,34 @@
...
+            "name": "Events to be emitted",
+            "type": "choice",
+            "hint": "The basic `heartbeat` event emitted by this plugin goes straight into the event ingestion queue. The optional `heartbeat_api` event goes via the event capture endpoint, which helps monitor the API too.",
+            "choices": [
+                "heartbeat",
+                "heartbeat + heartbeat_api"
+            ],
+            "default": "heartbeat",
+            "required": true
+        },
```
Comment (tiina303) [inline plugin.json]: Lets make this two separate boolean inputs that you can enable independently or both, making it a more user friendly experience :)

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/* global posthog */
+/* eslint no-undef: "error" */
+
+import fetch from 'node-fetch'
+
+export async function setupPlugin(meta) {
+    runEveryMinute(meta)
```
Comment (tiina303) [inline index.ts]: Why do we need this?

File: index.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+export async function runEveryMinute({ config }) {
+    const timestamp = new Date().toISOString()
```
Comment (tiina303) [inline index.ts:7]: can we do DateTime object so it's easier to compare (maybe DateTime in UTC)?

**Comment by tiina303** [comment]
test: https://github.com/PostHog/posthog-heartbeat-plugin/blob/main/package.json#L12 => index.ts

File: index.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    if (config.events.includes('heartbeat_api')) {
+        capturePromises.push(captureHeartbeatApi(timestamp, config.host, config.project_api_key))
+    }
+    await Promise.all(capturePromises)
+    console.log(`Captured: ${capturePromises.length}`)
+}
+
+/** Capture `heartbeat` event, directly into the queue */
+async function captureHeartbeat(timestamp: string) {
+    await posthog.capture('heartbeat', { $timestamp: timestamp })
```
Comment (tiina303) [inline index.ts:22]: should we use a different property name?

File: index.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+export async function runEveryMinute({ config }) {
+    const timestamp = new Date().toISOString()
+    const capturePromises = [captureHeartbeat(timestamp)]
+    if (config.events.includes('heartbeat_api')) {
+        capturePromises.push(captureHeartbeatApi(timestamp, config.host, config.project_api_key))
+    }
+    await Promise.all(capturePromises)
+    console.log(`Captured: ${capturePromises.length}`)
```
Comment (tiina303) [inline index.ts]: maybe console log debug timestamp too? Doesn't hurt to know when it was actually triggered in debug mode

File: plugin.json (hunk 1 above)
Comment (tiina303) [reply plugin.json]: discussed it's not an option to use checkboxes atm, though it would be nice if we extended the types

File: index.ts (hunk 2 above)
Comment (Twixes) [reply index.ts]: Oops, debug code

File: index.ts (hunk 5 above)
Comment (Twixes) [reply index.ts]: This was some initial debugging code but actually it wasn't very useful

**Review by tiina303 — APPROVED** [review]

**Pull request merged**